File: js/examples/gtpChat.ts
/**
 * Two-party GTP (text) chat — minimal in-memory end-to-end example.
 *
 * Demonstrates:
 *   - MLS two-party handshake
 *   - GBP node bootstrap
 *   - Sending text messages with CBOR and FlatBuffers codecs
 *   - Idempotency: duplicate (sender, messageId) returns status "duplicate"
 *
 * Run from repo root:
 *   npx ts-node js/examples/gtpChat.ts
 */

import { MlsContext } from "../src/mls";
import { GroupNode, StreamType } from "../src/node";
import { GtpClient, PayloadCodec } from "../src/gtp";

// --- MLS handshake ----------------------------------------------------------
const aliceMls = MlsContext.create("alice");
const bobMls   = MlsContext.create("bob");
bobMls.acceptWelcome(aliceMls.invite(bobMls.exportKeyPackage()));
console.log(`MLS epoch: alice=${aliceMls.epoch}  bob=${bobMls.epoch}`);

// --- GBP nodes --------------------------------------------------------------
const gid   = aliceMls.groupId;
const alice = GroupNode.create(1, gid);
const bob   = GroupNode.create(2, gid);
alice.bootstrapAsCreator(aliceMls.epoch);
bob.bootstrapAsJoiner(bobMls.epoch);

// --- GTP clients ------------------------------------------------------------
const gtpAlice = GtpClient.create();
const gtpBob   = GtpClient.create();

// Send "hello" with default CBOR codec.
const frame = gtpAlice.send(alice, aliceMls, 2, 1n, "hello");
for (const ev of bob.onWire(bobMls, frame.wire)) {
    if (ev.kind === "payload_received" && ev.streamType === StreamType.Text) {
        const r = gtpBob.accept(ev.plaintext!, bobMls.epoch, ev.codec ?? PayloadCodec.Cbor);
        console.log(`new (cbor):   text=${r.text}  status=${r.status}`);
    }
}

// Send with FlatBuffers codec.
const frame2 = gtpAlice.send(alice, aliceMls, 2, 2n, "hello flatbuffers", PayloadCodec.FlatBuffers);
for (const ev of bob.onWire(bobMls, frame2.wire)) {
    if (ev.kind === "payload_received" && ev.streamType === StreamType.Text) {
        const r = gtpBob.accept(ev.plaintext!, bobMls.epoch, ev.codec ?? PayloadCodec.Cbor);
        console.log(`new (fbs):    text=${r.text}  codec=${ev.codec}`);
    }
}

// Replay: same messageId=1n must come back as "duplicate".
const dup = gtpAlice.send(alice, aliceMls, 2, 1n, "hello");
for (const ev of bob.onWire(bobMls, dup.wire)) {
    if (ev.kind === "payload_received" && ev.streamType === StreamType.Text) {
        const r = gtpBob.accept(ev.plaintext!, bobMls.epoch);
        console.log(`replay:       status=${r.status}`);  // → duplicate
    }
}

// Cleanup.
[aliceMls, bobMls, alice, bob, gtpAlice, gtpBob].forEach(x => x.close());

File: crates/gbp/stack/examples/gtp_chat.rs
//! Two-party GTP (text) chat — minimal in-memory end-to-end example.
//!
//! Demonstrates:
//! - MLS two-party handshake
//! - GBP node bootstrap + epoch transition
//! - Sending text messages with CBOR and FlatBuffers codecs
//! - Idempotency: duplicate (sender, message_id) returns GtpAccept::Duplicate
//!
//! Run with: cargo run --example gtp_chat -p gbp-stack

use gbp_stack::{
    ControlOpcode, Event, GtpAccept, GtpClient, GroupNode, MlsContext, PayloadCodec, StreamType,
};
use openmls::prelude::{
    DeserializeBytes as _, KeyPackageIn, OpenMlsProvider as _, ProtocolVersion,
    tls_codec::Serialize as _,
};

fn main() -> anyhow::Result<()> {
    // --- MLS handshake -----------------------------------------------
    let (mut alice_mls, _) = MlsContext::new_member(b"alice")?;
    let (mut bob_mls, bob_kp_bundle) = MlsContext::new_member(b"bob")?;

    // Validate Bob's key package through Alice's crypto provider (openmls requirement).
    let bob_kp_bytes = bob_kp_bundle.key_package().tls_serialize_detached()?;
    let bob_kp = KeyPackageIn::tls_deserialize_exact_bytes(&bob_kp_bytes)?
        .validate(alice_mls.provider.crypto(), ProtocolVersion::Mls10)?;

    let (_commit, welcome) = alice_mls.invite_full(&[bob_kp])?;
    bob_mls.accept_welcome(&welcome)?;
    alice_mls.finalize_pending_commit()?;
    assert_eq!(alice_mls.epoch(), bob_mls.epoch());

    // --- GBP nodes ---------------------------------------------------
    let gid = alice_mls.group_id_16();
    let mut alice = GroupNode::new(1, gid);
    let mut bob   = GroupNode::new(2, gid);
    alice.bootstrap_as_creator(0);
    bob.bootstrap_as_joiner(0, 1);

    // Apply epoch-1 transition so both nodes are live.
    let exec = alice.send_control(
        &mut alice_mls, 0, ControlOpcode::ExecuteTransition, 1, 7, vec![],
    )?;
    alice.apply_transition(1);
    bob.on_wire(&mut bob_mls, &exec.wire)?;

    // --- GTP clients -------------------------------------------------
    let mut gtp_alice = GtpClient::new();
    let mut gtp_bob   = GtpClient::new();

    // Send "hello" with default CBOR codec.
    let frame = gtp_alice.send(&mut alice, &mut alice_mls, 2, 1, "hello", PayloadCodec::Cbor)?;
    for ev in bob.on_wire(&mut bob_mls, &frame.wire)? {
        if let Event::PayloadReceived(p) = ev {
            if p.stream_type == StreamType::Text {
                match gtp_bob.accept(&p.plaintext, bob_mls.epoch(), p.codec)? {
                    GtpAccept::New(msg)       => println!("new:       {:?}", msg.text()),
                    GtpAccept::Duplicate(msg) => println!("duplicate: {:?}", msg.text()),
                }
            }
        }
    }

    // Send again with FlatBuffers codec (lower decode overhead).
    let frame2 = gtp_alice.send(
        &mut alice, &mut alice_mls, 2, 2, "hello flatbuffers", PayloadCodec::FlatBuffers,
    )?;
    for ev in bob.on_wire(&mut bob_mls, &frame2.wire)? {
        if let Event::PayloadReceived(p) = ev {
            if p.stream_type == StreamType::Text {
                match gtp_bob.accept(&p.plaintext, bob_mls.epoch(), p.codec)? {
                    GtpAccept::New(msg) => println!("new (fbs): {:?}  codec={:?}", msg.text(), p.codec),
                    GtpAccept::Duplicate(msg) => println!("dup (fbs): {:?}", msg.text()),
                }
            }
        }
    }

    // Replay: resend message_id=1 — must come back as Duplicate.
    let dup = gtp_alice.send(&mut alice, &mut alice_mls, 2, 1, "hello", PayloadCodec::Cbor)?;
    for ev in bob.on_wire(&mut bob_mls, &dup.wire)? {
        if let Event::PayloadReceived(p) = ev {
            if p.stream_type == StreamType::Text {
                match gtp_bob.accept(&p.plaintext, bob_mls.epoch(), p.codec)? {
                    GtpAccept::New(_)       => println!("ERROR: expected duplicate"),
                    GtpAccept::Duplicate(_) => println!("replay correctly rejected as duplicate"),
                }
            }
        }
    }

    Ok(())
}
